packages/@aws-cdk/assertions/lib/match.ts

```diff
@@ -1,13 +1,242 @@
 import { ABSENT } from './vendored/assert';
 
 /**
- * Partial and special matching during template assertions
+ * Denotes a failure when matching patterns to targets.
  */
-export class Match {
+export interface MatchFailure {
+  /**
+   * The relative path at which the failure occurred.
+   */
+  readonly path: string[];
+  /**
+   * Description of the match failure.
+   */
+  readonly message: string;
+}
+
+/**
+ * Partial and special matching during template assertions.
+ */
+export abstract class Match {
   /**
    * Use this matcher in the place of a field's value, if the field must not be present.
    */
   public static absentProperty(): string {
     return ABSENT;
   }
+
+  /**
+   * Matches the specified pattern with the array found in the same relative path of the target.
+   * The set of elements (or matchers) must be in the same order as would be found.
+   * @param pattern the pattern to match
+   * @param options options to configure the matcher
+   */
+  public static arrayWith(pattern: any[], options: ArrayMatchOptions = {}): Match {
+    return new ArrayMatch(pattern, options);
+  }
+
+  /**
+   * Matches the specified pattern to an object found in the same relative path of the target.
+   * The keys and their values (or matchers) must be present in the target but the target can be a superset.
+   * @param pattern the pattern to match
+   * @param options options to configure the matcher
+   */
+  public static objectLike(pattern: {[key: string]: any}, options: ObjectMatchOptions = {}): Match {
+    return new ObjectMatch(pattern, options);
+  }
+
+  /**
+   * Check whether the provided object is a subtype of the `Match` class
+   */
+  public static isMatcher(x: any): x is Match {
+    return x && x instanceof Match;
+  }
+
+  /**
+   * Test whether a target matches the provided pattern.
+   * @param actual the target to match
+   * @return the list of match failures. An empty array denotes a successful match.
+   */
+  public abstract test(actual: any): MatchFailure[];
+}
+
+/**
+ * Options when initializing the `LiteralMatch` class.
+ */
+export interface LiteralMatchOptions {
+  /**
+   * Whether objects nested at any level should be matched partially.
+   * @default false
+   */
+  readonly partialObjects?: boolean;
+}
+
+/**
+ * A Match class that expects the target to match with the pattern exactly.
+ * The pattern may be nested with other matchers that are then deletegated to.
+ */
+export class LiteralMatch extends Match {
```

Comment:
Do we need to expose these?

Reply:
Not really, no. But there's a logistical challenge.

Since `Match` uses `XxxMatch` classes (in its static APIs) and `XxxMatch` inherits `Match`, they have to be in the same file. Otherwise, we get into import cycles.

`Match` should be exported publicly but `XxxMatch` exported internally for unit testing.

AFAIK, there are only two options here -
(a) export `XxxMatch` classes or,
(b) create another `Matcher` base class and continue to keep the static methods on `Match`.

I feel the former is the better option.
WDYT? Is there something I missed?

Reply:
`IMatcher` ?